dilation-protocol.md

```diff
@@ -1,371 +1,280 @@
+In all cases, the Leader is the only side which should send `reconnect`.
+
+```json
+{ "type": "reconnect" }
+```
+
+Upon receiving a `reconnect`, the Follower should stop any pending connection attempts and terminate any existing connections
+(even if they appear viable).
+Listening sockets may be retained, but any previous connection made through them must be dropped.
+
+Once all connections have stopped, the Follower should send a `reconnecting` message,
+then start the connection process for the next generation,
+which will send new `connection-hint` messages for all listening sockets.
+
+```json
+{ "type": "reconnecting" }
+```
+
+The Leader will drop all existing connections from before sending the `reconnect`,
+and will not initiate any new connections until it receives the matching `reconnecting` from the Follower.
+The Follower must drop all previous connections before it sends the `reconnecting` response.
+
+~~(TODO: what about a follower->leader connection that was started before
 start-dilation is received, and gets established on the Leader side after
 start-dilation is sent? the follower will drop it after it receives
-start-dilation, but meanwhile the leader may accept it as gen2)
+start-dilation, but meanwhile the leader may accept it as gen2 ~warner)~~
 
 (probably need to include the generation number in the handshake, or in the
-derived key)
-
-(TODO: reduce the number of round-trip stalls here, I've added too many)
-
-Each side is in the "connecting" state (which encompasses both making
-connection attempts and having an established connection) starting with the
-receipt of a `please` message and a local `w.dilate()` call. The
-Leader remains in that state until it abandons the connection and sends a
-`reconnect` message, at which point it remains in the "flushing" state until
-the Follower's `reconnecting` message is received. The Follower remains in
-"connecting" until it receives `reconnect`, then it stays in "dropping" until
-it finishes halting all outstanding connections, after which it sends
-`reconnecting` and switches back to "connecting".
-
-"Connection hints" are type/address/port records that tell the other side of
-likely targets for L2 connections. Both sides will try to determine their
-external IP addresses, listen on a TCP port, and advertise `(tcp,
-external-IP, port)` as a connection hint. The Transit Relay is also used as a
-(lower-priority) hint. These are sent in `connection-hint` records, which can
-be sent any time after both sending and receiving a `please` record. Each
-side will initiate connections upon receipt of the hints.
+derived key ~warner — please please don't ~piegames)
 
-```
-{ "type": "connection-hints",
-  "hints": [ ... ]
-}
-```
+(TODO: reduce the number of round-trip stalls here, I've added too many ~warner)
+(yes, and one could also drastically reduce the number of messages by sending
+all hints at once with the reconnect message ~piegames)
```

Comment:
> There would need to be at least two separate ones: "not-relayed" and "relayed".

Transit manages to do this while sending all hints at once, by delaying the connection attempt towards relay servers instead. This allows more flexibility in the implementation: For example, the Rust code connects to everything, but when a connection is a relay connection it waits a couple of seconds to see whether a "better" connection comes up.

> I think adding this requirement will slow the protocol down, because then you can't send any message until all possible transports are set up. Tor transports (for example) can takes "dozens of seconds" to establish.

This could be solved by booting up these connections up early (we already know whether the other side supports it as soon as we get their `versions` message).

Reply:
The current spec _allows_ you to do this, doesn't it? (I mean: a "hints" message can already contain multiple hints). We shouldn't be too prescriptive here, there's just too many things going on and many, many possible network scenarios.